### PR TITLE
[SVS2] discrete svs development 

### PR DESCRIPTION
**Pull request by A-Quarter-Mile**
The latest updates of svs2 will be listed here.
SVS2 aims to utilize discrete tokens replace Mel spectrograms as intermediate representation.

For now, the first version of svs2 recipe is ready for the Mandarin dataset Opencpop.

- [x] S3PRL
- [x] MERT
- [x] EnCodec
- [ ] ...

Details will be supplemented.

### CODE REVIEW

**Comment by sw005320** [comment]
@simpleoier, can you review this PR?

File: egs2/ofuton_p_utagoe_db/svs2/run.sh
Comment (sw005320) [inline egs2/ofuton_p_utagoe_db/svs2/run.sh:1]: There are too many arguments. Can you reduce them by removing the default values (or making some of them default values in svs2 template), etc.?

File: egs2/opencpop/svs2/run.sh
Comment (sw005320) [inline egs2/opencpop/svs2/run.sh:1]: Ditto. Too many arguments.

**Review by sw005320 — COMMENTED** [review]
The PR looks good to me.
Please make sure to add a test and prepare a document.

File: egs2/ofuton_p_utagoe_db/svs2/local/customed_dic.scp
Comment (simpleoier) [inline egs2/ofuton_p_utagoe_db/svs2/local/customed_dic.scp:1]: This could be a soft link from svs1. Either is fine.

File: egs2/ofuton_p_utagoe_db/svs2/local/data.sh
Comment (simpleoier) [inline egs2/ofuton_p_utagoe_db/svs2/local/data.sh:1]: ditto?

File: egs2/ofuton_p_utagoe_db/svs2/local/dataset_split.py
Comment (simpleoier) [inline egs2/ofuton_p_utagoe_db/svs2/local/dataset_split.py:1]: ditto?

File: egs2/ofuton_p_utagoe_db/svs2/local/prep_segments.py
Comment (simpleoier) [inline egs2/ofuton_p_utagoe_db/svs2/local/prep_segments.py:1]: ditto?

File: egs2/ofuton_p_utagoe_db/svs2/run.sh
Comment (simpleoier) [reply egs2/ofuton_p_utagoe_db/svs2/run.sh:1]: I agree with Shinji.

File: egs2/tedlium3/asr2/README.md
Comment (simpleoier) [inline egs2/tedlium3/asr2/README.md:1]: Why is this file in the current PR? `egs2/tedlium3/asr2`

File: egs2/tedlium3/asr2/asr2.sh
Comment (simpleoier) [inline egs2/tedlium3/asr2/asr2.sh:1]: ditto

File: egs2/tedlium3/asr2/cmd.sh
Comment (simpleoier) [inline egs2/tedlium3/asr2/cmd.sh:1]: ditto.

Comment (simpleoier) [reply egs2/tedlium3/asr2/cmd.sh:1]: And the following files.

File: egs2/wsj0_2mix/enh1/README.md
Comment (simpleoier) [inline egs2/wsj0_2mix/enh1/README.md:1]: And`egs2/wsj0_2mix/enh1`?

File: espnet2/enh/separator/tfgridnetv2_separator.py
Comment (simpleoier) [inline espnet2/enh/separator/tfgridnetv2_separator.py:1]: `espnet2/enh/separator`?

File: espnet2/svs/discrete/loss.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+        """
+        super(DiscreteLoss, self).__init__()
+        assert (use_masking != use_weighted_masking) or not use_masking
+        self.use_masking = use_masking
+        self.use_weighted_masking = use_weighted_masking
+
+        # define criterions
+        reduction = "none" if self.use_weighted_masking else "mean"
+        # self.l1_criterion = torch.nn.L1Loss(reduction=reduction)
```
Comment (simpleoier) [inline espnet2/svs/discrete/loss.py:54]: clean this?

File: espnet2/tasks/enh.py
Comment (simpleoier) [inline espnet2/tasks/enh.py:1]: `espnet2/tasks/enh.py`?

File: test/espnet2/enh/test_espnet_model.py
Comment (simpleoier) [inline test/espnet2/enh/test_espnet_model.py:1]: Is this PR related to enh tasks?

**Review by simpleoier — COMMENTED** [review]
Thanks for your efforts! I left some comments.

File: egs2/TEMPLATE/svs2/svs2.sh [hunk 2]
```diff
@@ -0,0 +1,1654 @@
...
+
+		utt_extra_files="${utt_extra_files} utt2sid"
+            done
+        fi
+    fi
+
+
+    if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then
+        if "${use_lid}"; then
+            log "Stage 2+: lid extract: data/ -> ${data_feats}/"
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply egs2/TEMPLATE/svs2/svs2.sh:536]: Agree, please differentiate them @A-Quarter-Mile

File: egs2/TEMPLATE/svs2/svs2.sh [hunk 3]
```diff
@@ -0,0 +1,1654 @@
...
+                    "data/${dset}/wav.scp" "${data_feats}${_suf}/${dset}"
+                scripts/audio/format_score_scp.sh --nj "${nj}" --cmd "${train_cmd}" \
+                    ${_opts} \
+                    "data/${dset}/score.scp" "${data_feats}${_suf}/${dset}"
+                echo "${feats_type}" > "${data_feats}${_suf}/${dset}/feats_type"
+            done
+        fi
+    fi
+    # TODO(Yuning): Introducing a single new stage for conditional token generation
+    if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then
```
[the remark this answers is not in the document]
Comment (ftshijt) [reply egs2/TEMPLATE/svs2/svs2.sh:505]: As per my understanding, it is reserved for potential embedding vector generation. As it usually needs additional checks/revalidating, merging it with format wavscp sometimes making it difficult to debug/verify etc.

But yeah, the stage index should change regarding to that.

**Comment by ftshijt** [comment]
Hi @A-Quarter-Mile , since we are developing on the branch first, next time, you may not need to update the your local fork with the espent/master branch as it would introduce extra burdens for our reviews. We can focus more on the conflicts when we finalize the PR and prepare for merging to master branch.

**Comment by ftshijt** [comment]
@A-Quarter-Mile Please do not merge master directly for this PR (as you are merging to svs2 branch, merging with master makes the PR difficult to read)

**Comment by ftshijt** [comment]
Also, please let us know if the PR is ready for review

File: egs2/ofuton_p_utagoe_db/svs2/local/data.sh [hunk 4]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env bash
```
Comment (ftshijt) [inline egs2/ofuton_p_utagoe_db/svs2/local/data.sh:1]: Please use symlink for the same files (similar to other files)

**Comment by ftshijt** [comment]
Given that the initial version is working, I'm thinking of merging the PR as the initial step. Could you please let us know if the initial updates are ready for review? @A-Quarter-Mile @South-Twilight

File: espnet2/bin/svs_inference.py [hunk 5]
```diff
@@ -326,6 +346,8 @@ def inference(
     vocoder_config: Optional[str] = None,
     vocoder_checkpoint: Optional[str] = None,
     vocoder_tag: Optional[str] = None,
+    discrete_token_layers: int = 1,
+    mix_type: str = "frame",
```
Comment (ftshijt) [inline espnet2/bin/svs_inference.py:350]: Can  you explain a bit more on what is this for?

File: egs2/opencpop/svs2/run.sh [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+#    --src_lang ${src_lang} \
+#    --tgt_lang ${tgt_lang} \
+#    --src_token_type "char" \
+#    --src_nbpe $src_nbpe \
+#    --tgt_token_type "char" \
+#    --tgt_nbpe $tgt_nbpe \
+#    --src_case ${src_case} \
+#    --tgt_case ${tgt_case} \
+#    --inference_config "${inference_config}" \
+#    --lm_train_text "data/${train_set}/text.${tgt_case}.${tgt_lang} data/local/other_text/text" "$@"
```
Comment (ftshijt) [inline egs2/opencpop/svs2/run.sh:83]:
```suggestion
```

File: egs2/opencpop/svs2/run.sh [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+fs=16000
+fmin=80
+fmax=7600
+n_fft=2048
+n_shift=320
+win_length=1200
+score_feats_extract=syllable_score_feats   # frame_score_feats | syllable_score_feats
+
+kmeans_feature="mfcc"
+# "wavlm_large/6" | "encodec/1" | "xls_r_300m/6" # use model_type/layer_index
```
Comment (ftshijt) [inline egs2/opencpop/svs2/run.sh:18]: maybe we can put the supported list to readme? (maybe in next PR)

File: egs2/opencpop/svs2/local/midi-note.scp [hunk 8]
```diff
@@ -0,0 +1,107 @@
+127	G9
```
Comment (ftshijt) [inline egs2/opencpop/svs2/local/midi-note.scp:1]: let's use the existing one

File: egs2/opencpop/svs2/local/data_prep.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
+import argparse
```
Comment (ftshijt) [inline egs2/opencpop/svs2/local/data_prep.py:1]: Can we reuse the existing one?

File: egs2/opencpop/svs2/local/data.sh [hunk 10]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env bash
```
Comment (ftshijt) [inline egs2/opencpop/svs2/local/data.sh:1]: Can we use the existing one?

File: egs2/opencpop/svs2/conf/tuning/train_xiaoice.yaml [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+optim: adam           # optimizer type
+optim_conf:           # keyword arguments for selected optimizer
+    lr: 1.0e-03       # learning rate
+    eps: 1.0e-06      # epsilon
+    weight_decay: 0.0 # weight decay coefficient
+
+##########################################################
+#                OTHER TRAINING SETTING                  #
+##########################################################
+#num_iters_per_epoch: 500    # number of iterations per epoch
```
Comment (ftshijt) [inline egs2/opencpop/svs2/conf/tuning/train_xiaoice.yaml:53]:
```suggestion
```

File: egs2/opencpop/svs2/README.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
+# SVS2: Discrete Singing Voice Synthesis
+
+We build discrete token based SVS systems.
+Details to be updated.
+
+## Results
+We list results from different models on svs2.
+
+| Token type                                             | MCD | f0   | wavs   |
```
Comment (ftshijt) [inline egs2/opencpop/svs2/README.md:9]:
```suggestion
| Token type                                             | MCD | f0rmse   | wavs   |
```

File: egs2/ofuton_p_utagoe_db/svs2/run.sh [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    --src_lang ${src_lang} \
+    --tgt_lang ${tgt_lang} \
+    --src_token_type "char" \
+    --src_nbpe $src_nbpe \
+    --tgt_token_type "char" \
+    --tgt_nbpe $tgt_nbpe \
+    --src_case ${src_case} \
+    --tgt_case ${tgt_case} \
+    --inference_config "${inference_config}" \
+    --lm_train_text "data/${train_set}/text.${tgt_case}.${tgt_lang} data/local/other_text/text" "$@"
```
Comment (ftshijt) [inline egs2/ofuton_p_utagoe_db/svs2/run.sh:82]: Do we need these?

File: egs2/ofuton_p_utagoe_db/svs2/local/dataset_split.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
+import argparse
```
Comment (ftshijt) [inline egs2/ofuton_p_utagoe_db/svs2/local/dataset_split.py:1]: Please reuse if possible

File: egs2/TEMPLATE/svs2/svs2.sh [hunk 15]
```diff
@@ -0,0 +1,1445 @@
...
+        kmeans_feature_conf="{type=mert,conf={fs=24000,multilayer_feature=False,layer=${layer},download_path=${mert_url}}}"
+    elif [ ${kmeans_feature_type} = "encodec" ]; then
+        kmeans_feature_conf="{type=encodec,conf={fs=48000,bandwidth=12,multilayer_feature=False,layer=${layer},download_path=${encodec_url}}}"
+    elif [ ${kmeans_feature_type} != "multi" ]; then
+        s3prl_conf="{upstream=${kmeans_feature_type}}"
+        kmeans_feature_conf="{type=s3prl,conf={s3prl_conf=${s3prl_conf},download_dir=ckpt,multilayer_feature=False,layer=${layer}}}"
+    fi
+fi
+if [ ${kmeans_feature_type} = "multi" ]; then
+    token_name=${layer}
```
Comment (ftshijt) [inline egs2/TEMPLATE/svs2/svs2.sh:306]: Do we have an example somewhere for the multi-token examples?

File: egs2/TEMPLATE/svs2/svs2.sh [hunk 16]
```diff
@@ -0,0 +1,1445 @@
...
+                --km_dir "${km_dir}" \
+                --portion "${portion}" \
+                --nclusters "${nclusters}" \
+                --storage_save_mode ${storage_save_mode} \
+                --use_gpu true \
+                --nj ${nj} \
+                --cpu_cmd "${train_cmd}" \
+                --cuda_cmd "${cuda_cmd}" \
+                ${kmeans_opts}
+        fi
```
Comment (ftshijt) [inline egs2/TEMPLATE/svs2/svs2.sh:600]: We seems missing the multi-option in this stage?

**Review by ftshijt — COMMENTED** [review]
Thanks for the update! Could you please address the comments (including the previous comments from other reviewers)

**Comment by ftshijt** [comment]
Hi @A-Quarter-Mile, could you please respond to the previous comments, so that we can merge the PR?

**Comment by ftshijt** [comment]
@A-Quarter-Mile  Any chances to fix this PR?
